Permission Model for MongoDB

khoaguin opened the pull request:
Related issues:
[[Security] Enable Permissions Model in MongoDB Store#642](https://github.com/OpenMined/Heartbeat/issues/642)
[Inconsistency in Display Order with different Databases#645](https://github.com/OpenMined/Heartbeat/issues/645)

khoaguin committed:
Type checking for the `permission` argument of `has_permission` of MongoStorePartition

khoaguin committed:
Merge branch 'dev' into mongodb-permission-model

khoaguin committed:
Merge branch 'dev' into mongodb-permission-model

khoaguin committed:
Merge branch 'dev' into mongodb-permission-model

khoaguin committed:
add `with_collection_permissions` function that create the collections that store the permissions

khoaguin committed:
Merge branch 'dev' into mongodb-permission-model

khoaguin committed:
specify output type

khoaguin committed:
add "READ" permission to User and NodeSettings permissions collections in `_set` method of MongoStorePartition

khoaguin committed:
Merge with 'dev'

khoaguin commented:
Now after `node = sy.orchestra.launch(name="test-domain", reset=True)` we have 2 permissions collections for `User` and `NodeSettings`
![image](https://github.com/OpenMined/PySyft/assets/88959106/8186e04d-dddb-4349-ab6f-1531ec41d5f2)

khoaguin committed:
Get back permissions and check permission in `MongoStorePartition.has_permission()`

khoaguin committed:
Merge with 'dev'

khoaguin committed:
implementing `add_permission`

Co-authored-by: Shubham Gupta <shubhamgupta3121@gmail.com>

khoaguin committed:
done implementing `add_permission`

khoaguin committed:
done the `remove_permission` method of MongoStorePartition

khoaguin committed:
Merge w 'dev'

khoaguin committed:
implementing `has_permission`

khoaguin committed:
done implementing `take_ownership` for MongoStorePartition

shubham3121 committed:
minor fixes to object permission at mongo partition store

khoaguin committed:
Merge with 'dev'

khoaguin committed:
fix a small issue while passing credentials to ActionObjectPermission.
Try calling `take_ownership` of the MongoStorePartition to make it similar to KeyValueStorePartition

shubham3121 committed:
minor fixes

shubham3121 committed:
Merge branch 'dev' into mongodb-permission-model

shubham3121 reviewed (approved):
Wonderful work @khoaguin !! 🎉 🔥

koenvanderveen commented on packages/syft/src/syft/store/mongo_document_store.py:
```diff
@@ -211,23 +232,43 @@ def _set(
+        collection = collection_status.ok()
+
+        store_key_exists = (
+            collection.find_one({"_id": store_query_key.value}) is not None
```
can we assume this is always an id?

khoaguin replied:
Actually we can't be sure since `store_query_key` is a `QueryKey` and its `value` has type `Any`
![image](https://github.com/OpenMined/PySyft/assets/88959106/3496ce0b-ad80-441e-a10b-2e271a917f47)

shubham3121 replied:
`store_query_key.as_dict_mongo` handles that logic. Updated the logic there.
Resolved here: https://github.com/OpenMined/PySyft/pull/8017/commits/6b5cc8e70c46e6564e53b2d988a774b2030b7f21

koenvanderveen commented on packages/syft/src/syft/store/mongo_document_store.py:
```diff
@@ -211,23 +232,43 @@ def _set(
                 collection.insert_one(storage_obj)
             except DuplicateKeyError as e:
                 return Err(f"Duplicate Key Error for {obj}: {e}")
+
```
not sure if the logic here is completely the same as the logic in the kv_docstore, (althought that logic also seems a bit flawed). In the kv_doc store, if the item already exists and we do ignore duplicates, we would **not write**, but return the original object instead. I am not saying that that logic makes sense, but it looks like there is a discrepancy between that and this logic

khoaguin replied:
when not setting `WriteConcern(w=0)`, an error will be raised if the item already exists.
When we set  `WriteConcern(w=0))`, adding existed items will not raise error, and the database will not be changed (so it kept the existed item)

koenvanderveen replied:
I resolved this in the latest commits, please have a look @khoaguin @shubham3121

koenvanderveen commented on packages/syft/src/syft/store/mongo_document_store.py:
```diff
@@ -345,12 +386,117 @@ def _delete(
         return Err(f"Failed to delete object with qk: {qk}")
 
     def has_permission(self, permission: ActionObjectPermission) -> bool:
```
:+1:

koenvanderveen commented:
Overall this PR looks very good to me, great work @khoaguin! I am not completely sure about the test coverage of the mongo store, do we have any visibility of on which tests use this? Also, it would be quite good to have a notebook that shows how you can easily start a node with a mongo store using orchestra (maybe for a later PR as well, but we can make a ticket now).

khoaguin commented:
Thank you @koenvanderveen . From what I know we have the tests in `mongo_document_store_test.py` but they are more about the functionality of the methods like `set`, `update`..., rather than the permissions

koenvanderveen commented:
> Thank you @koenvanderveen . From what I know we have the tests in `mongo_document_store_test.py` but they are more about the functionality of the methods like `set`, `update`..., rather than the permissions

Oke, I think its not necessary to do it now, but we can at least add a ticket for triaging

khoaguin commented:
Yes sure @koenvanderveen. Please start the ticket and assign it to me, I think I can and should start working on it right away since memory is still fresh. By the way, I am done with the notebook to test this feature. Where should I put it?

khoaguin committed:
Merge branch 'dev' into mongodb-permission-model

shubham3121 committed:
Merge branch 'dev' into mongodb-permission-model

khoaguin commented:
Hi @shubham3121  @koenvanderveen, here is the notebook containing some code to test the feature
[mongodb-permission-model.zip](https://github.com/OpenMined/PySyft/files/12496379/mongodb-permission-model.zip)

koenvanderveen committed:
-

khoaguin committed:
add type `MongoCollection` for the `collection` variable for easier read and debug)

khoaguin committed:
Merge branch 'mongodb-permission-model' of https://github.com/khoaguin/PySyft into mongodb-permission-model

khoaguin committed:
Merge branch 'dev' into mongodb-permission-model

shubham3121 committed:
update to use store_query_key.as_dict_mongo

shubham3121 commented:
@koenvanderveen @khoaguin We're running testing the notebooks against mongo in the CI. Check pr-tests-notebook-stack which runs `tox -e stack.test.notebook` with `ORCHESTRA_DEPLOYMENT_TYPE: "container_stack"`, which launches docker containers with Mongo. Task runner: `om-ci-16vcpu-ubuntu2204`

koenvanderveen committed:
check for uniqueness instead of relying on mongo to ignore errors

koenvanderveen committed:
Merge branch 'mongodb-permission-model' of github.com:khoaguin/PySyft into mongodb-permission-model

shubham3121 commented on packages/syft/src/syft/store/mongo_document_store.py:
```diff
@@ -211,27 +231,60 @@ def _set(
+            can_write: bool = ownership_result.is_ok()
+        elif not ignore_duplicates:
+            unique_query_keys: QueryKeys = self.settings.unique_keys.with_obj(obj)
+            keys = ", ".join(f"`{key.key}`" for key in unique_query_keys.all)
```
I'm fine with this for now, but this is a redundant check as we already create unique indexes in Mongo against unique keys (`_create_update_index`), which should already take care of this logic.

khoaguin commented:
After the commit [a914c87](https://github.com/OpenMined/PySyft/pull/8017/commits/a914c87fe7ba2ddf0eb43463fc95baf4b20144df), we had a bug like in the picture below where we got error if we add things with `ignore_duplicates = True` in `_set()` of `MongoStorePartition`, for example when using `data_subject_registry.add_data_subject`

![image](https://github.com/OpenMined/PySyft/assets/88959106/a0f47488-b70a-4721-8596-9b4ed5d1d019)

I reverted back to the commit [6b5cc8e](https://github.com/OpenMined/PySyft/pull/8017/commits/6b5cc8e70c46e6564e53b2d988a774b2030b7f21) where we did not get this issue like below,
![image](https://github.com/OpenMined/PySyft/assets/88959106/96655cad-aa66-445a-ac88-84dc0ded6159),
however, we still got an issue: Although we only had 3 data subjects
![image](https://github.com/OpenMined/PySyft/assets/88959106/d4ab3ca7-68f2-4ddc-895a-c7588eec684b)
but in the `DataSubject_permissions` collection we have 4 entries like below
![image](https://github.com/OpenMined/PySyft/assets/88959106/b8b48626-7919-4b70-a876-b60fdd078e50)

Furthermore, if I tried to re-upload the same dataset to create duplicates, we will get an error when uploading the dataset
![image](https://github.com/OpenMined/PySyft/assets/88959106/6740ef39-d459-4327-964f-fbbf2a1de170)
and will still only have 1 dataset. However, we added 1 more permission to the `Dataset_permissions` collection
![image](https://github.com/OpenMined/PySyft/assets/88959106/da0f94a6-083a-4138-b583-63edb81b6df5)

kiendang committed:
Merge branch 'dev' into mongodb-permission-model

shubham3121 committed:
fix item_keys_exist query to be OR instead of AND

khoaguin commented:
Recheck after this commit [a9d6290](https://github.com/OpenMined/PySyft/pull/8017/commits/a9d6290995a8273100a9267cfcb5725b7ab66749) and previous issues within [this comment](https://github.com/OpenMined/PySyft/pull/8017#issuecomment-1704720674) are fixed
